File: chile-open-data/data_sources/scripts/download_example.py
#!/usr/bin/env python3
"""
Script de descarga de ejemplo - Fase 1
Biblioteca de Datos Abiertos de Chile

Este script demuestra cómo descargar datasets usando la configuración
definida en sources.yaml.
"""

import yaml
import requests
import os
from pathlib import Path
from typing import Dict, List
import argparse
import sys

def load_sources(sources_path: Path) -> List[Dict]:
    """Carga la configuración de fuentes desde sources.yaml"""
    if not sources_path.exists():
        raise FileNotFoundError(f"No se encontró {sources_path}")
    
    with open(sources_path, 'r', encoding='utf-8') as f:
        data = yaml.safe_load(f)
    
    # Soporta tanto 'datasets' como 'sources'
    datasets = data.get('datasets', data.get('sources', []))
    if not datasets:
        raise ValueError("No se encontraron datasets en la configuración")
    
    return datasets

def check_dataset_availability(dataset: Dict) -> Dict:
    """Verifica si un dataset está disponible"""
    url = dataset['url']
    method = dataset.get('method', 'HEAD').upper()
    timeout = dataset.get('timeout', 10)
    
    try:
        if method == 'HEAD':
            response = requests.head(url, timeout=timeout, allow_redirects=True)
        else:
            response = requests.get(url, timeout=timeout, stream=True)
        
        return {
            'id': dataset['id'],
            'name': dataset['name'],
            'url': url,
            'status': 'available' if response.status_code < 400 else 'unavailable',
            'status_code': response.status_code,
            'category': dataset['category']
        }
    except requests.RequestException as e:
        return {
            'id': dataset['id'],
            'name': dataset['name'],
            'url': url,
            'status': 'error',
            'error': str(e),
            'category': dataset['category']
        }

def download_dataset(dataset: Dict, output_dir: Path, dry_run: bool = False) -> bool:
    """
    Descarga un dataset (simulado en esta versión de ejemplo)
    En una implementación completa, aquí se haría la descarga real
    """
    output_dir.mkdir(parents=True, exist_ok=True)
    
    if dry_run:
        print(f"[DRY-RUN] Descargaría: {dataset['name']} -> {output_dir}")
        return True
    
    # En una implementación real, aquí iría la lógica de descarga específica
    # por tipo de dataset (CSV, API, scraping, etc.)
    print(f"⚠️  Descarga simulada: {dataset['name']}")
    print(f"   URL: {dataset['url']}")
    print(f"   Categoría: {dataset['category']}")
    
    # Crear un archivo de ejemplo
    example_file = output_dir / f"{dataset['id']}_metadata.txt"
    with open(example_file, 'w', encoding='utf-8') as f:
        f.write(f"Dataset: {dataset['name']}\n")
        f.write(f"URL: {dataset['url']}\n")
        f.write(f"Categoría: {dataset['category']}\n")
        f.write(f"Descripción: {dataset.get('description', 'Sin descripción')}\n")
    
    print(f"   Metadata guardada en: {example_file}")
    return True

def main():
    parser = argparse.ArgumentParser(description='Descarga datasets de Chile')
    parser.add_argument('--sources', '-s', 
                       default='data_sources/config/sources.yaml',
                       help='Ruta al archivo sources.yaml')
    parser.add_argument('--output', '-o',
                       default='downloads',
                       help='Directorio de salida')
    parser.add_argument('--dataset', '-d',
                       help='ID específico de dataset a descargar')
    parser.add_argument('--check-only', '-c',
                       action='store_true',
                       help='Solo verificar disponibilidad, no descargar')
    parser.add_argument('--dry-run',
                       action='store_true',
                       help='Simular descarga sin ejecutar')
    parser.add_argument('--category',
                       help='Filtrar por categoría específica')
    
    args = parser.parse_args()
    
    # Configurar rutas
    sources_path = Path(args.sources)
    output_dir = Path(args.output)
    
    try:
        # Cargar configuración
        datasets = load_sources(sources_path)
        print(f"📂 Cargados {len(datasets)} datasets desde {sources_path}")
        
        # Filtrar por dataset específico
        if args.dataset:
            datasets = [ds for ds in datasets if ds['id'] == args.dataset]
            if not datasets:
                print(f"❌ Dataset '{args.dataset}' no encontrado")
                sys.exit(1)
        
        # Filtrar por categoría
        if args.category:
            datasets = [ds for ds in datasets if ds['category'] == args.category]
            if not datasets:
                print(f"❌ No se encontraron datasets en la categoría '{args.category}'")
                sys.exit(1)
        
        # Verificar disponibilidad
        print("\n🔍 Verificando disponibilidad de datasets...")
        results = []
        for dataset in datasets:
            result = check_dataset_availability(dataset)
            results.append(result)
            
            status_emoji = "✅" if result['status'] == 'available' else "❌"
            print(f"{status_emoji} {result['name']} ({result['category']}) - {result['status']}")
        
        if args.check_only:
            print(f"\n📊 Resumen: {len([r for r in results if r['status'] == 'available'])} disponibles de {len(results)}")
            return
        
        # Descargar datasets disponibles
        available_datasets = [ds for ds, result in zip(datasets, results) 
                            if result['status'] == 'available']
        
        if not available_datasets:
            print("\n❌ No hay datasets disponibles para descargar")
            return
        
        print(f"\n⬇️  Descargando {len(available_datasets)} datasets...")
        success_count = 0
        
        for dataset in available_datasets:
            dataset_dir = output_dir / dataset['category'] / dataset['id']
            if download_dataset(dataset, dataset_dir, args.dry_run):
                success_count += 1
        
        print(f"\n✅ Proceso completado: {success_count}/{len(available_datasets)} datasets procesados")
        
    except Exception as e:
        print(f"❌ Error: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: chile-open-data/data_sources/scripts/download_example.R
# Script de descarga de ejemplo - Fase 1
# Biblioteca de Datos Abiertos de Chile
# 
# Este script demuestra cómo descargar datasets usando la configuración
# definida en sources.yaml desde R.

# Cargar librerías necesarias
if (!require(yaml)) install.packages("yaml")
if (!require(httr)) install.packages("httr")
if (!require(jsonlite)) install.packages("jsonlite")

library(yaml)
library(httr)
library(jsonlite)

# Función para cargar configuración de fuentes
load_sources <- function(sources_path = "data_sources/config/sources.yaml") {
  if (!file.exists(sources_path)) {
    stop(paste("No se encontró", sources_path))
  }
  
  data <- read_yaml(sources_path)
  
  # Soporta tanto 'datasets' como 'sources'
  datasets <- data$datasets
  if (is.null(datasets)) {
    datasets <- data$sources
  }
  
  if (is.null(datasets) || length(datasets) == 0) {
    stop("No se encontraron datasets en la configuración")
  }
  
  return(datasets)
}

# Función para verificar disponibilidad de un dataset
check_dataset_availability <- function(dataset) {
  url <- dataset$url
  method <- ifelse(is.null(dataset$method), "HEAD", toupper(dataset$method))
  timeout_val <- ifelse(is.null(dataset$timeout), 10, as.numeric(dataset$timeout))
  
  tryCatch({
    if (method == "HEAD") {
      response <- HEAD(url, timeout(timeout_val))
    } else {
      response <- GET(url, timeout(timeout_val))
    }
    
    list(
      id = dataset$id,
      name = dataset$name,
      url = url,
      status = ifelse(status_code(response) < 400, "available", "unavailable"),
      status_code = status_code(response),
      category = dataset$category
    )
  }, error = function(e) {
    list(
      id = dataset$id,
      name = dataset$name,
      url = url,
      status = "error",
      error = as.character(e),
      category = dataset$category
    )
  })
}

# Función para descargar un dataset (simulado)
download_dataset <- function(dataset, output_dir = "downloads", dry_run = FALSE) {
  # Crear directorio de salida
  dataset_dir <- file.path(output_dir, dataset$category, dataset$id)
  if (!dir.exists(dataset_dir)) {
    dir.create(dataset_dir, recursive = TRUE)
  }
  
  if (dry_run) {
    cat(sprintf("[DRY-RUN] Descargaría: %s -> %s\n", dataset$name, dataset_dir))
    return(TRUE)
  }
  
  # En una implementación real, aquí iría la lógica de descarga específica
  cat(sprintf("⚠️  Descarga simulada: %s\n", dataset$name))
  cat(sprintf("   URL: %s\n", dataset$url))
  cat(sprintf("   Categoría: %s\n", dataset$category))
  
  # Crear un archivo de ejemplo
  metadata_file <- file.path(dataset_dir, paste0(dataset$id, "_metadata.txt"))
  metadata_content <- paste(
    sprintf("Dataset: %s", dataset$name),
    sprintf("URL: %s", dataset$url),
    sprintf("Categoría: %s", dataset$category),
    sprintf("Descripción: %s", ifelse(is.null(dataset$description), "Sin descripción", dataset$description)),
    sep = "\n"
  )
  
  writeLines(metadata_content, metadata_file)
  cat(sprintf("   Metadata guardada en: %s\n", metadata_file))
  
  return(TRUE)
}

# Función principal de descarga
download_chile_data <- function(sources_path = "data_sources/config/sources.yaml",
                               output_dir = "downloads",
                               dataset_id = NULL,
                               category = NULL,
                               check_only = FALSE,
                               dry_run = FALSE) {
  
  tryCatch({
    # Cargar configuración
    datasets <- load_sources(sources_path)
    cat(sprintf("📂 Cargados %d datasets desde %s\n", length(datasets), sources_path))
    
    # Filtrar por dataset específico
    if (!is.null(dataset_id)) {
      datasets <- datasets[sapply(datasets, function(ds) ds$id == dataset_id)]
      if (length(datasets) == 0) {
        stop(sprintf("Dataset '%s' no encontrado", dataset_id))
      }
    }
    
    # Filtrar por categoría
    if (!is.null(category)) {
      datasets <- datasets[sapply(datasets, function(ds) ds$category == category)]
      if (length(datasets) == 0) {
        stop(sprintf("No se encontraron datasets en la categoría '%s'", category))
      }
    }
    
    # Verificar disponibilidad
    cat("\n🔍 Verificando disponibilidad de datasets...\n")
    results <- lapply(datasets, check_dataset_availability)
    
    for (result in results) {
      status_emoji <- ifelse(result$status == "available", "✅", "❌")
      cat(sprintf("%s %s (%s) - %s\n", status_emoji, result$name, result$category, result$status))
    }
    
    if (check_only) {
      available_count <- sum(sapply(results, function(r) r$status == "available"))
      cat(sprintf("\n📊 Resumen: %d disponibles de %d\n", available_count, length(results)))
      return(invisible(results))
    }
    
    # Descargar datasets disponibles
    available_indices <- which(sapply(results, function(r) r$status == "available"))
    available_datasets <- datasets[available_indices]
    
    if (length(available_datasets) == 0) {
      cat("\n❌ No hay datasets disponibles para descargar\n")
      return(invisible(results))
    }
    
    cat(sprintf("\n⬇️  Descargando %d datasets...\n", length(available_datasets)))
    success_count <- 0
    
    for (dataset in available_datasets) {
      if (download_dataset(dataset, output_dir, dry_run)) {
        success_count <- success_count + 1
      }
    }
    
    cat(sprintf("\n✅ Proceso completado: %d/%d datasets procesados\n", 
                success_count, length(available_datasets)))
    
    return(invisible(results))
    
  }, error = function(e) {
    cat(sprintf("❌ Error: %s\n", as.character(e)))
    stop(e)
  })
}

# Ejemplos de uso:

# Verificar disponibilidad de todos los datasets
# download_chile_data(check_only = TRUE)

# Descargar todos los datasets disponibles (simulado)
# download_chile_data(dry_run = TRUE)

# Descargar solo datasets de economía
# download_chile_data(category = "economía", dry_run = TRUE)

# Descargar un dataset específico
# download_chile_data(dataset_id = "bcentral_pib", dry_run = TRUE)

# Función auxiliar para mostrar ayuda
show_help <- function() {
  cat("📖 Biblioteca de Datos Abiertos de Chile - Script R\n\n")
  cat("Funciones disponibles:\n")
  cat("• download_chile_data(): Función principal de descarga\n")
  cat("• load_sources(): Cargar configuración desde YAML\n")
  cat("• check_dataset_availability(): Verificar un dataset\n\n")
  cat("Parámetros principales:\n")
  cat("• sources_path: Ruta al archivo sources.yaml\n")
  cat("• output_dir: Directorio de descarga\n")
  cat("• dataset_id: ID específico de dataset\n")
  cat("• category: Filtrar por categoría\n")
  cat("• check_only: Solo verificar disponibilidad\n")
  cat("• dry_run: Simular descarga\n\n")
  cat("Ejemplos:\n")
  cat("download_chile_data(check_only = TRUE)\n")
  cat("download_chile_data(category = 'economía', dry_run = TRUE)\n")
}

# Mostrar ayuda al cargar el script
cat("✅ Script de descarga R cargado correctamente.\n")
cat("💡 Usa show_help() para ver ejemplos de uso.\n")
